components/micro-gateway-core/src/main/ballerina/src/gateway/jwt_generator/jwt_gen_util.bal

```diff
@@ -217,3 +191,89 @@ public function loadJWTGeneratorImpl() returns boolean {
+# + cacheKey - key for the jwt generator cache
+# + enabledCaching - jwt generator caching enabled
+# + return - Returns `true` if the token generation and setting the header completed successfully
+public function setJWTHeader(BackendJWTGenUserContextDTO tokenContextDTO,
+                            map<string> apiDetails,
+                            http:Request req,
+                            string cacheKey,
+                            boolean enabledCaching) returns @tainted boolean {
+
+    (handle|error) generatedToken;
+    jwt:JwtPayload? payload = tokenContextDTO.payload;
+    if (payload is jwt:JwtPayload) {
+        if (isSelfContainedToken(payload)) {
+            generatedToken = generateJWTToken(payload, apiDetails);
+            return setGeneratedTokenAsHeader(req, cacheKey, enabledCaching, generatedToken);
+        }
+    }
+    ClaimsMapDTO claimsMapDTO = createMapFromRetrievedUserClaimsListDTO(tokenContextDTO);
+    generatedToken = generateJWTTokenFromUserClaimsMap(claimsMapDTO, apiDetails);
+    if (generatedToken is error) {
+        printError(KEY_JWT_AUTH_PROVIDER, "Token not generated due to error", generatedToken);
+        return false;
+    }
+    return setGeneratedTokenAsHeader(req, cacheKey, enabledCaching, generatedToken);
+}
+
+# Set the BackendJWT header after checking in the cache.
+# The JWT token is picked from cache if it is available in the cache and the time duration between current time
+# and the cached token's expiry time is less than the skew time.
+# If the token is not picked from the cache, then it will be generated.
+#
+# + req - HTTP Request
+# + cacheKey - Cache key used for jwt generator cache
+# + enabledCaching - True if caching is enabled for backend JWT generation
+# + tokenContextDTO - User Context which is used to populate the information required for jwtGenerator Implementation
+# + apiDetails - API related information which is used to populate the information required for jwtGenerator Implementation
+# + return - Returns `true` if adding the JWT token to the request is successful.
+function setJWTTokenWithCacheCheck(http:Request req,
+                                    string cacheKey,
+                                    int skewTime,
+                                    boolean enabledCaching,
+                                    BackendJWTGenUserContextDTO tokenContextDTO,
+                                    map<string> apiDetails)
+                                    returns @tainted boolean {
+    boolean status = false;
+    if (enabledCaching) {
+        var cachedToken = jwtGeneratorCache.get(cacheKey);
+        printDebug(KEY_JWT_AUTH_PROVIDER, "Key: " + cacheKey);
+        if (cachedToken is string) {
+            printDebug(KEY_JWT_AUTH_PROVIDER, "Found in jwt generator cache");
+            printDebug(KEY_JWT_AUTH_PROVIDER, "Token: " + cachedToken);
+            (jwt:JwtPayload | error) cachedPayload = getDecodedJWTPayload(cachedToken, tokenContextDTO.issuer);
```

Comment:
Since you only want to decode the jwt here(not checking in the jwt issuer's cache), use the following method. 
```suggestion
            (jwt:JwtPayload | error) cachedPayload = decodeJWTPayload(cachedToken);
```

Reply:
If I call decodeJWTPayload it will try to decode all the time. But in this case, it will only decode if the token is not there in cache.

Reply:
fixed